[JSC] GreedyRegAlloc: Rewrite forEachTmpInGroup to use a worklist rather than recursion

dhecht opened the pull request:
#### c1ef250810d99963a1bcf5a50b79666ade015fc5
<pre>
[JSC] GreedyRegAlloc: Rewrite forEachTmpInGroup to use a worklist rather than recursion
<a href="https://bugs.webkit.org/show_bug.cgi?id=294366">https://bugs.webkit.org/show_bug.cgi?id=294366</a>
<a href="https://rdar.apple.com/153014929">rdar://153014929</a>

Reviewed by Yusuke Suzuki and Keith Miller.

Avoid overflowing the stack by using a worklist tree traversal rather
than recursive. The worklist is owned by the caller to avoid
unnecessary allocator churn.

Canonical link: <a href="https://commits.webkit.org/296147@main">https://commits.webkit.org/296147@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d3c8f8201cdca1d0d59d4053ce7ce7d9ff22d482

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107485 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27167 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17575 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112699 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58021 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27852 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35668 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81594 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110414 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22060 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96870 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61975 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21497 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15011 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57465 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/100074 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91426 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15042 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115800 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/106032 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34552 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25458 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90627 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34927 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93126 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90373 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23051 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35287 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13063 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30299 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34473 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40019 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/130348 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34244 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35448 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37574 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35880 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -1158,6 +1167,7 @@ class GreedyAllocator {
             }
         };
         Vector<Move> moves;
+        Vector<Tmp> worklist0, worklist1;
```
Maybe we should have some inlineCapacity?

dhecht replied:
Sure, done.

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -1466,9 +1476,10 @@ class GreedyAllocator {
 
         ScalarRegisterSet alreadyAttempted;
         if (eagerGroupsExhaustiveSearch) {
+            Vector<Tmp> worklist;
```
Ditto.

Constellation commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -1132,15 +1132,24 @@ class GreedyAllocator {
+
+            if (data.isGroup()) {
+                worklist.append(data.subGroup0);
+                worklist.append(data.subGroup1);
```
Previously we are handling subGroup0 first, but now, it is reversed, is it intentional?

dhecht replied:
The tree isn't ordered but I've switched it in case that's less surprising.

kmiller68 reviewed (approved):
r=me too.

dhecht committed:
[JSC] GreedyRegAlloc: Rewrite forEachTmpInGroup to use a worklist rather than recursion
https://bugs.webkit.org/show_bug.cgi?id=294366
rdar://153014929

Reviewed by Yusuke Suzuki and Keith Miller.

Avoid overflowing the stack by using a worklist tree traversal rather
than recursive. The worklist is owned by the caller to avoid
unnecessary allocator churn.

Canonical link: https://commits.webkit.org/296147@main
